### ISSUE TITLE
Android Camera2 not working

### ISSUE THREAD

**Issue by Dakror** [opencv/opencv#11325]
##### System information (version)

- OpenCV => 3.4.1
- Operating System / Platform => Android 5.0.1
- Compiler => -/-

##### Detailed description

The Camera2 API Classes Camera2Renderer and JavaCamera2View are not functioning correctly. At least on my phone (Huawei P8, Android 5.0.1). JavaCamera2View crashes in Line 265 while accessing sizes[0] of the possible camera sizes with an IndexOutOfBoundsException 0 >= 0, and Camera2Renderer, used through CameraGLSurfaceView results in a preview with a distorted aspect ratio. The Android Samples are using obsolete and outdated code that doesn't even compile with this version of openCV.

### Minimal example

 ```java
import android.app.Activity;
import android.os.Bundle;
import android.view.View;
import android.view.WindowManager;

import org.opencv.android.CameraGLSurfaceView;

public class MainActivity extends Activity {
    static {
        System.loadLibrary("opencv_java3");
    }

    private CameraGLSurfaceView mView;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mView = new CameraGLSurfaceView(this, null);
        setContentView(mView);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
            WindowManager.LayoutParams.FLAG_FULLSCREEN);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

        View decorView = getWindow().getDecorView();

        int uiOptions = View.SYSTEM_UI_FLAG_HIDE_NAVIGATION | View.SYSTEM_UI_FLAG_FULLSCREEN;
        decorView.setSystemUiVisibility(uiOptions);

    }

    @Override
    protected void onPause() {
        mView.onPause();
        super.onPause();
    }

    @Override
    protected void onResume() {
        super.onResume();
        mView.onResume();
    }
}
```

### Results 

holding phone in landscape
![screenshot_2018-04-16-21-32-49](https://user-images.githubusercontent.com/2452821/38830923-144b0c50-41be-11e8-94e5-f4404837107a.png)

in portrait
![screenshot_2018-04-16-21-32-37](https://user-images.githubusercontent.com/2452821/38830927-16489d9c-41be-11e8-9235-11b7597c0ea3.png)

### Minimal example
```java
import android.app.Activity;
import android.os.Bundle;
import android.view.View;
import android.view.WindowManager;

import org.opencv.android.JavaCamera2View;

public class MainActivity extends Activity {
    static {
        System.loadLibrary("opencv_java3");
    }

    private JavaCamera2View mView;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);

        mView = findViewById(R.id.camera);
        mView.enableView();

        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
            WindowManager.LayoutParams.FLAG_FULLSCREEN);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

        View decorView = getWindow().getDecorView();

        int uiOptions = View.SYSTEM_UI_FLAG_HIDE_NAVIGATION | View.SYSTEM_UI_FLAG_FULLSCREEN;
        decorView.setSystemUiVisibility(uiOptions);

    }
}
```

layout/activity_main.xml
```xml
<?xml version="1.0" encoding="utf-8"?>
<LinearLayout
    xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto"
    xmlns:opencv="http://schemas.android.com/apk/res-auto"
    xmlns:tools="http://schemas.android.com/tools"
    android:layout_width="match_parent"
    android:layout_height="match_parent"
    tools:context=".MainActivity">

<org.opencv.android.JavaCamera2View
    android:layout_width="match_parent"
    android:id="@+id/camera"
    android:layout_height="match_parent"/>

</LinearLayout>
```

### Stacktrace

```
    java.lang.RuntimeException: Interrupted while setCameraPreviewSize.
        at org.opencv.android.JavaCamera2View.connectCamera(JavaCamera2View.java:320)
        at org.opencv.android.CameraBridgeViewBase.onEnterStartedState(CameraBridgeViewBase.java:360)
        at org.opencv.android.CameraBridgeViewBase.processEnterState(CameraBridgeViewBase.java:321)
        at org.opencv.android.CameraBridgeViewBase.checkCurrentState(CameraBridgeViewBase.java:313)
        at org.opencv.android.CameraBridgeViewBase.surfaceChanged(CameraBridgeViewBase.java:198)
        at android.view.SurfaceView.updateWindow(SurfaceView.java:614)
        at android.view.SurfaceView$3.onPreDraw(SurfaceView.java:187)
        at android.view.ViewTreeObserver.dispatchOnPreDraw(ViewTreeObserver.java:847)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2018)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1094)
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6054)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:793)
        at android.view.Choreographer.doCallbacks(Choreographer.java:606)
        at android.view.Choreographer.doFrame(Choreographer.java:575)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:779)
        at android.os.Handler.handleCallback(Handler.java:739)
        at android.os.Handler.dispatchMessage(Handler.java:95)
        at android.os.Looper.loop(Looper.java:135)
        at android.app.ActivityThread.main(ActivityThread.java:5538)
        at java.lang.reflect.Method.invoke(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:372)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:958)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:753)
     Caused by: java.lang.ArrayIndexOutOfBoundsException: length=0; index=0
        at org.opencv.android.JavaCamera2View.calcPreviewSize(JavaCamera2View.java:265)
        at org.opencv.android.JavaCamera2View.connectCamera(JavaCamera2View.java:300)
        at org.opencv.android.CameraBridgeViewBase.onEnterStartedState(CameraBridgeViewBase.java:360) 
        at org.opencv.android.CameraBridgeViewBase.processEnterState(CameraBridgeViewBase.java:321) 
        at org.opencv.android.CameraBridgeViewBase.checkCurrentState(CameraBridgeViewBase.java:313) 
        at org.opencv.android.CameraBridgeViewBase.surfaceChanged(CameraBridgeViewBase.java:198) 
        at android.view.SurfaceView.updateWindow(SurfaceView.java:614) 
        at android.view.SurfaceView$3.onPreDraw(SurfaceView.java:187) 
        at android.view.ViewTreeObserver.dispatchOnPreDraw(ViewTreeObserver.java:847) 
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2018) 
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1094) 
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6054) 
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:793) 
        at android.view.Choreographer.doCallbacks(Choreographer.java:606) 
        at android.view.Choreographer.doFrame(Choreographer.java:575) 
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:779) 
        at android.os.Handler.handleCallback(Handler.java:739) 
        at android.os.Handler.dispatchMessage(Handler.java:95) 
        at android.os.Looper.loop(Looper.java:135) 
        at android.app.ActivityThread.main(ActivityThread.java:5538) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:372) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:958) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:753)
```

---

The only thing i got to work on my phone, is JavaCameraView, but only at low 15 FPS.

**Comment by floe** [comment]
Just found this issue - note that although the Camera2 API was introduced in Android 5, most implementations before Android 7 (or even 8) are very buggy. If `sizes[0] == null`, then this is most likely an issue in the Camera2 implementation itself.

**Comment by Zulqurnain** [comment]
I'm getting the same error what is the solution?

**Comment by alalek** [comment]
should be fixed by #13840
